docs/ARCHITECTURE.md

```diff
@@ -0,0 +1,113 @@
+# ARCHITECTURE
+
+An `ActiveModel::Serializer` is essentially a [Decorator](http://c2.com/cgi/wiki?DecoratorPattern)
+for a resource. It allows you to specify which attributes and
+associations should represent the serialization of the resource. It requires an adapter to
+transform its attributes into a JSON document; it cannot be serialized itself.
+
+The `ActiveModel::ArraySerializer` represent a collection of resources as serializers
+and, if there is no serializer, primitives.
+
+The `ActiveModel::Adapter` describes the structure of the JSON document generated from a
+serializer. For example, the `Attributes` of example represents each serializer as its
+unmodified attributes.  The `JsonApi` adatper represents the serializer as a [JSON
+API](jsonapi.org/) document.
+
+The `ActiveModel::SerializableResource` acts to coordinate the serializer(s) and adapter
+to an object that responds to `to_json`, and `as_json`.  It is used in the controller to
+encapsulate the serialization resource when rendered. Thus, it can be used on its own
+to serialize a resource outside of a controller, as well.
+
+## Primitive handling
+
+Definitions: A primitive is usually a String or Array. There is no serializer
+defined for them; they will be serialized when the resource is converted to JSON (`as_json` or
+`to_json`).  (The below also applies for any object with no serializer.)
+
+ActiveModelSerializers doesn't handle primitives passed to `render json:` at all.
+
+However, when a primitive value is an attribute or in a collection,
+it is is not modified.
+
+Internally, if no serializer can be found in the controller, the resource is not decorated by
+ActiveModelSerializers.
+
+If the collection serializer (ArraySerializer) cannot
+identify a serializer for a resource in its collection, it raises [`NoSerializerError`](https://github.com/rails-api/active_model_serializers/issues/1191#issuecomment-142327128)
+which is rescued in `AcitveModel::Serializer::Reflection#build_association` which sets
+the association value directly:
+
+```ruby
+reflection_options[:virtual_value] = association_value.try(:as_json) || association_value
+```
+
+(which is called by the adapter as `serializer.associations(*)`.)
+
+## How options are parsed
+
+High-level overview:
+
+- For a collection
+  - the collection serializer is the `:serializer` option and
+  - `:each_serializer` is used as the serializer for each resource in the collection.
+- For a single resource, the `:serializer` option is the resource serializer.
+- All options go to the serializer except those specified for the adapter as `ADAPTER_OPTIONS`.
```

Comment:
All options go to the *serializer* may be a bit ambiguous here, can you clarify?

Also, `ADAPTER_OPTIONS` hasn't been previously mentioned, so that may lead to some confusion.